Optimize reduction + amax fusion

ipiszy opened the pull request:
This PR optimizes cases like layer_norm + fp8 quant (which includes amax and fp8 quant) fusion when amax is split into multiple reduction kernels.

Benchmark:
```
python test/inductor/test_fp8.py -k test_layernorm_fp8_quant_benchmark

Before this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.13262102689486555ms, Eager: 0.8211962616822429ms, LN only Inductor: 0.09606276150627614ms.

After this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.08281274131274131ms, Eager: 0.8217452830188678ms, LN only Inductor: 0.09586902286902287ms.
```

LN + fp8 quant is even faster than LN itself. The reason could be that LN + fp8 outputs fp8 while LN outputs fp16.

From Inductor nightly benchmark test:
There are perf differences in cuda_graph / cuda_graph_dynamic / default runs, but no difference in inductor_max_autotune. So it seems to me that the perf differences are mostly like fluctuations. 

![Screenshot 2023-10-18 at 4 58 55 PM](https://github.com/pytorch/pytorch/assets/10527447/6640474a-1e1d-4d33-97e9-0a60d0bc9f1f)



Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111122



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

ipiszy committed:
Optimize reduction + amax fusion

[ghstack-poisoned]

ipiszy commented on torch/_inductor/dependencies.py:
```diff
@@ -359,5 +359,44 @@ def extract_read_writes(
+        # Input node has already been realized. Return its size and reduction_size.
+        return input_node.get_size(), input_node.get_reduction_size()
+
+    # This is one issue: what if there are permutations between the input node and its dependent realized nodes?
```
@jansel Wonder do you have any suggestions for this?

jansel replied:
In addition to permutations there are views which change the ndimension.

Is it ok if this function is approximate?  Or are there correctness issues if it is wrong?

ipiszy replied:
Using reduction_sizes from dependent nodes have a better chance to fuse these nodes.

e.g. The current case is:
```
x1 = layer_norm(x0)
x2 = amax(x1)
x3 = to_fp8(x1)
```

Inductor generates these nodes:
```
n0=WelfordReduction()
n1=WelfordReduction()
n2=WelfordReduction()
n3=Pointwise()
n4=Reduction()
n5=Pointwise()
```

Currently n0, n1, n2, n3, n5 are fused together. n3, n4 are fused together.
I'd like to make first level reduction ranges of n4 the same as n0 / n1 / n2, so that n0, n1, n2, n3, first level n4, n5 can be fused together.

So it seem to me that we cannot use approximate values here for n4 reduction sizes.

vadimkantorov commented on test/inductor/test_fp8.py:
```diff
@@ -98,6 +121,183 @@ def fp8_cast(x, dtype):
+        batch_size, sequence_length, hidden_size = shape
+
+        def amax_fp8(x: Tensor, scale: Tensor):
+            y = torch.max(torch.abs(x))
```
should this be `amax` instead of `max`? If `max` is not intentional, I think using `amax` to mean "return the values without indices" is clearer

ipiszy committed:
Update on "Optimize reduction + amax fusion"

This PR optimizes cases like layer_norm + fp8 quant (which includes amax and fp8 quant) fusion when amax is split into multiple reduction kernels.

Benchmark:
```
python test/inductor/test_fp8.py -k test_layernorm_fp8_quant_benchmark

Before this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.13262102689486555ms, Eager: 0.8211962616822429ms, LN only Inductor: 0.09606276150627614ms.

After this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.08281274131274131ms, Eager: 0.8217452830188678ms, LN only Inductor: 0.09586902286902287ms.
```

LN + fp8 quant is even faster than LN itself. The reason could be that LN + fp8 outputs fp8 while LN outputs fp16.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ipiszy committed:
Update on "Optimize reduction + amax fusion"

This PR optimizes cases like layer_norm + fp8 quant (which includes amax and fp8 quant) fusion when amax is split into multiple reduction kernels.

Benchmark:
```
python test/inductor/test_fp8.py -k test_layernorm_fp8_quant_benchmark

Before this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.13262102689486555ms, Eager: 0.8211962616822429ms, LN only Inductor: 0.09606276150627614ms.

After this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.08281274131274131ms, Eager: 0.8217452830188678ms, LN only Inductor: 0.09586902286902287ms.
```

LN + fp8 quant is even faster than LN itself. The reason could be that LN + fp8 outputs fp8 while LN outputs fp16.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ipiszy committed:
Update on "Optimize reduction + amax fusion"

This PR optimizes cases like layer_norm + fp8 quant (which includes amax and fp8 quant) fusion when amax is split into multiple reduction kernels.

Benchmark:
```
python test/inductor/test_fp8.py -k test_layernorm_fp8_quant_benchmark

Before this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.13262102689486555ms, Eager: 0.8211962616822429ms, LN only Inductor: 0.09606276150627614ms.

After this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.08281274131274131ms, Eager: 0.8217452830188678ms, LN only Inductor: 0.09586902286902287ms.
```

LN + fp8 quant is even faster than LN itself. The reason could be that LN + fp8 outputs fp8 while LN outputs fp16.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ipiszy committed:
Update on "Optimize reduction + amax fusion"

This PR optimizes cases like layer_norm + fp8 quant (which includes amax and fp8 quant) fusion when amax is split into multiple reduction kernels.

Benchmark:
```
python test/inductor/test_fp8.py -k test_layernorm_fp8_quant_benchmark

Before this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.13262102689486555ms, Eager: 0.8211962616822429ms, LN only Inductor: 0.09606276150627614ms.

After this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.08281274131274131ms, Eager: 0.8217452830188678ms, LN only Inductor: 0.09586902286902287ms.
```

LN + fp8 quant is even faster than LN itself. The reason could be that LN + fp8 outputs fp8 while LN outputs fp16.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on torch/_inductor/dependencies.py:
```diff
@@ -359,5 +359,50 @@ def extract_read_writes(
+
+    from .ir import ComputedBuffer, Loops
+
+    if not isinstance(input_node.data.data, Loops):
```
I think we need some checks to ensure `.data` and `.data.data` exist.  There are some cases like views that result in different nesting.

jansel commented on torch/_inductor/dependencies.py:
```diff
@@ -359,5 +359,50 @@ def extract_read_writes(
+
+    if not isinstance(input_node.data.data, Loops):
+        # Input node has already been realized. Return its size and reduction_size.
+        if hasattr(input_node, "get_size") and hasattr(
+            input_node, "get_reduction_size"
+        ):
```
Adding a method would be cleaner than these hasattr checks.

ipiszy committed:
Update on "Optimize reduction + amax fusion"

This PR optimizes cases like layer_norm + fp8 quant (which includes amax and fp8 quant) fusion when amax is split into multiple reduction kernels.

Benchmark:
```
python test/inductor/test_fp8.py -k test_layernorm_fp8_quant_benchmark

Before this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.13262102689486555ms, Eager: 0.8211962616822429ms, LN only Inductor: 0.09606276150627614ms.

After this PR:
Config: float8_dtype=torch.float8_e5m2, shape=(4, 2048, 4096). 
Benchmark results: Inductor: 0.08281274131274131ms, Eager: 0.8217452830188678ms, LN only Inductor: 0.09586902286902287ms.
```

LN + fp8 quant is even faster than LN itself. The reason could be that LN + fp8 outputs fp8 while LN outputs fp16.

From Inductor nightly benchmark test:
There are perf differences in cuda_graph / cuda_graph_dynamic / default runs, but no difference in inductor_max_autotune. So it seems to me that the perf differences are mostly like fluctuations. 

![Screenshot 2023-10-18 at 4 58 55 PM](https://github.com/pytorch/pytorch/assets/10527447/6640474a-1e1d-4d33-97e9-0a60d0bc9f1f)






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ipiszy reviewed (commented):
Thanks @jansel !

ipiszy commented:
@pytorchbot merge

ipiszy commented:
@pytorchbot label "topic: not user facing"

ipiszy commented:
@pytorchbot merge

mlazos commented:
@ipiszy This PR caused a [significant regression in TIMM dm_nfnet_f0](https://hud.pytorch.org/benchmark/timm_models/inductor_with_cudagraphs?startTime=Mon,%2016%20Oct%202023%2023:53:25%20GMT&stopTime=Mon,%2023%20Oct%202023%2023:53:25%20GMT&granularity=hour&mode=training&model=dm_nfnet_f0&dtype=amp&lBranch=main&lCommit=3ca81aed4261f7bce3111c8905d979872f545f12&rBranch=main&rCommit=90e2117a99e935bbe38ce4493bad85960d5cda6e)

repro command:

`python timm_models.py --training --amp --performance --only=dm_nfnet_f0 --inductor`

Can you take a look?

cc @eellison

ipiszy commented:
> @ipiszy This PR caused a [significant regression in TIMM dm_nfnet_f0](https://hud.pytorch.org/benchmark/timm_models/inductor_with_cudagraphs?startTime=Mon,%2016%20Oct%202023%2023:53:25%20GMT&stopTime=Mon,%2023%20Oct%202023%2023:53:25%20GMT&granularity=hour&mode=training&model=dm_nfnet_f0&dtype=amp&lBranch=main&lCommit=3ca81aed4261f7bce3111c8905d979872f545f12&rBranch=main&rCommit=90e2117a99e935bbe38ce4493bad85960d5cda6e)
> 
> repro command:
> 
> `python timm_models.py --training --amp --performance --only=dm_nfnet_f0 --inductor`
> 
> Can you take a look?
> 
> cc @eellison

FYI this is fixed by https://github.com/pytorch/pytorch/pull/112297.
